### PR TITLE
Diofant import:  Improve coverage status

### PR DESCRIPTION
**Pull request by Upabjojr**
Port PR:
https://github.com/diofant/diofant/pull/222/files

### CODE REVIEW

File: sympy/core/add.py [hunk 1]
```diff
@@ -437,12 +436,6 @@ def as_numer_denom(self):
         for f in expr.args:
             ni, di = f.as_numer_denom()
             nd[di].append(ni)
-        # put infinity in the numerator
```
Comment (smichr) [inline sympy/core/add.py]: what does `(1 + Pow(0,-1,evaluate=False)).as_numer_denom()` with this removal?

File: sympy/series/order.py [hunk 2]
```diff
@@ -385,6 +385,8 @@ def contains(self, expr):
             for s in common_symbols:
                 l = ratio.limit(s, point)
                 from sympy.series.limits import Limit
+
+                l = Limit(ratio, s, point).doit(heuristics=False)
```
Comment (smichr) [inline sympy/series/order.py]: What is a test that can be added that will differ in result if this line is not here?

File: sympy/core/add.py (hunk 1 above)
Comment (Upabjojr) [reply sympy/core/add.py]: `(zoo, 1)`, same as in the latest release.

File: sympy/series/order.py (hunk 2 above)
Comment (Upabjojr) [reply sympy/series/order.py]: At first sight, it doesn't look like there's a difference. Maybe @skirpichev had some wider plans in mind. Anyway, _diofant_ appears to be much faster than _SymPy_ at calculating limits, so I think it's worth porting @skirpichev 's improvements back to _SymPy_.

Comment (skirpichev) [reply sympy/series/order.py]: > At first sight, it doesn't look like there's a difference.

There is.  Now O.contains doesn't use heuristics, but algorithm.

> diofant appears to be much faster than SymPy at calculating limits

If you just worry about speed, I bet - sympy dev's could add a lot of new hacks which make your current code faster than diofant's.

File: sympy/series/order.py [hunk 3]
```diff
@@ -385,6 +385,8 @@ def contains(self, expr):
             for s in common_symbols:
                 l = ratio.limit(s, point)
```
Comment (skirpichev) [inline sympy/series/order.py]: Dead code.

**Comment by smichr** [comment]
porting @skirpichev 's improvements back to SymPy.

I have reservations about this in light of the differences in licences. While removal of code seems fair, modifications that mirror those of diofant which were not arrived at independently  by identifying the issue and then coming up with a solution don't seem to be something that we should be doing. Am I misunderstanding the license differences?

**Comment by Upabjojr** [comment]
> While removal of code seems fair, modifications that mirror those of diofant which were not arrived at independently by identifying the issue and then coming up with a solution don't seem to be something that we should be doing.

Well, I have been trying to port some of the edits in diofant that could be useful to us. For example, diofant now uses `Pow` to represent `exp`. Unfortunately the limits section has changed a lot.

Diofant's gruntz algorithm looks much cleaner than SymPy's (we would definitely benefit from porting that code) and apparently even faster.

The problem of importing the code from diofant is that there are so many edits over the last years that having a one-time import could be a huge fuss.

> Am I misunderstanding the license differences?

We already include diofant's license into SymPy's license, so we can import the code provided that @skirpichev stays the author of the commits that he has written.

**Comment by skirpichev** [comment]
> provided that @skirpichev stays the author of the commits that he has written

You must add reference to the commit source (specific commit, PR, or at least - Diofant's repo itself).  Authorship less important, as I understand.

**Comment by Upabjojr** [comment]
> You must add reference to the commit source (specific commit, PR, or at least - Diofant's repo itself). 

I'm using `git cherry-pick` as you suggested in the other PR. Is that not enough?

**Comment by asmeurer** [comment]
Would it be possible to amend the commit message for each commit with the original diofant commit hash?

**Comment by ylemkimon** [comment]
With `git cherry-pick` option `-x`, it automatically appends `(cherry picked from commit [original hash])`, and with `-e`, you can edit the commit message.

**Comment by Upabjojr** [comment]
I think this can be merged.

**Pull request merged**